src/resolve/remote.rs

```diff
@@ -0,0 +1,141 @@
+use std::fs::{File, OpenOptions, create_dir_all, remove_file};
+use std::io;
+use std::path::{Path, PathBuf};
+
+// TODO: The error representation is awkward with reqwest, evaluate cHTTP instead.
+use reqwest::{Client, Error as RequestError, RedirectPolicy, Response, header};
+use url::Url;
+
+/// Provide access to remotely hosted resources.
+pub struct Remote {
+    client: Client,
+    temp: PathBuf,
+}
+
+pub struct Downloaded {
+    file: File,
+    path: PathBuf,
+    content_type: Option<ContentType>,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Request(RequestError),
+    Io(io::Error),
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum ContentType {
+    Markdown,
+    Image,
+    Pdf,
+}
+
+impl Remote {
+    pub fn new(download_folder: PathBuf) -> Result<Self, Error> {
+        create_dir_all(&download_folder)?;
+
+        let client = Client::builder()
+            // TODO: how should redirects interact relative references etc. ?
+            .redirect(RedirectPolicy::none())
+            .build()?;
+
+        Ok(Remote {
+            temp: download_folder,
+            client,
+        })
+    }
+
+    pub fn http(&self, url: Url) -> Result<Downloaded, Error> {
+        let mut response = self.client.get(url.as_ref())
+            // TODO set headers: user agent, accepted content type, ...
+            .send()
+            .and_then(|response| response.error_for_status())?;
+
+        let path = self.target_path(&url);
+
+        create_dir_all(path.parent().unwrap())?;
+
+        // Replace whatever file already existed.
+        // TODO: proper caching
+        let _ = remove_file(&path);
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(&path)?;
+
+        io::copy(&mut response, &mut file)?;
+
+        let content_type = self.content_type(&response);
+
+        Ok(Downloaded {
+            file,
+            path,
+            content_type,
+        })
+    }
+
+    fn content_type(&self, response: &Response) -> Option<ContentType> {
+        // Get a value content type header if any.
+        let header = response.headers().get(header::CONTENT_TYPE);
+        let header = header.and_then(|value| value.to_str().ok());
+
+        match header {
+            Some("text/markdown") => Some(ContentType::Markdown),
+            Some("image/png") | Some("image/jpeg") => Some(ContentType::Image),
+            Some("application/pdf") => Some(ContentType::Pdf),
```

Comment:
Can we use the `mime` crate here?

Reply:
In some first impression, the `mime` crate complicates this match expression. But it could help by also allowing media types with added parameters such as text encoding.